File: packages/backend/dynamodb.py
"""
DynamoDB table setup — run once to create local tables.
Usage: python dynamodb.py
"""
import os
import boto3
from decimal import Decimal
from datetime import datetime, timezone
from dotenv import load_dotenv

load_dotenv(".env.local")

ENDPOINT = os.getenv("DYNAMODB_ENDPOINT")
REGION = os.getenv("AWS_REGION", "us-east-1")


def get_client():
    kwargs = {"region_name": REGION}
    if ENDPOINT:
        kwargs["endpoint_url"] = ENDPOINT
        kwargs["aws_access_key_id"] = "local"
        kwargs["aws_secret_access_key"] = "local"
    return boto3.client("dynamodb", **kwargs)


def create_tables():
    client = get_client()
    existing = client.list_tables()["TableNames"]

    tables = [
        {
            "TableName": "finops_runs",
            "KeySchema": [
                {"AttributeName": "run_id", "KeyType": "HASH"},
                {"AttributeName": "sk", "KeyType": "RANGE"},
            ],
            "AttributeDefinitions": [
                {"AttributeName": "run_id", "AttributeType": "S"},
                {"AttributeName": "sk", "AttributeType": "S"},
                {"AttributeName": "status", "AttributeType": "S"},
                {"AttributeName": "started_at", "AttributeType": "S"},
            ],
            "GlobalSecondaryIndexes": [
                {
                    "IndexName": "status-started_at-index",
                    "KeySchema": [
                        {"AttributeName": "status", "KeyType": "HASH"},
                        {"AttributeName": "started_at", "KeyType": "RANGE"},
                    ],
                    "Projection": {"ProjectionType": "ALL"},
                    "ProvisionedThroughput": {"ReadCapacityUnits": 5, "WriteCapacityUnits": 5},
                }
            ],
            "BillingMode": "PAY_PER_REQUEST",
        },
        {
            "TableName": "finops_host_lists",
            "KeySchema": [
                {"AttributeName": "tenant_id", "KeyType": "HASH"},
                {"AttributeName": "run_id", "KeyType": "RANGE"},
            ],
            "AttributeDefinitions": [
                {"AttributeName": "tenant_id", "AttributeType": "S"},
                {"AttributeName": "run_id", "AttributeType": "S"},
            ],
            "BillingMode": "PAY_PER_REQUEST",
        },
        {
            "TableName": "finops_org_summary",
            "KeySchema": [
                {"AttributeName": "tenant_id", "KeyType": "HASH"},
                {"AttributeName": "run_id", "KeyType": "RANGE"},
            ],
            "AttributeDefinitions": [
                {"AttributeName": "tenant_id", "AttributeType": "S"},
                {"AttributeName": "run_id", "AttributeType": "S"},
            ],
            "GlobalSecondaryIndexes": [
                {
                    "IndexName": "run_id-index",
                    "KeySchema": [
                        {"AttributeName": "run_id", "KeyType": "HASH"},
                    ],
                    "Projection": {"ProjectionType": "ALL"},
                    "ProvisionedThroughput": {"ReadCapacityUnits": 5, "WriteCapacityUnits": 5},
                }
            ],
            "BillingMode": "PAY_PER_REQUEST",
        },
        {
            "TableName": "finops_host_results",
            "KeySchema": [
                {"AttributeName": "tenant_id", "KeyType": "HASH"},
                {"AttributeName": "sk", "KeyType": "RANGE"},
            ],
            "AttributeDefinitions": [
                {"AttributeName": "tenant_id", "AttributeType": "S"},
                {"AttributeName": "sk", "AttributeType": "S"},
                {"AttributeName": "run_id", "AttributeType": "S"},
            ],
            "GlobalSecondaryIndexes": [
                {
                    "IndexName": "run_id-index",
                    "KeySchema": [
                        {"AttributeName": "run_id", "KeyType": "HASH"},
                    ],
                    "Projection": {"ProjectionType": "ALL"},
                    "ProvisionedThroughput": {"ReadCapacityUnits": 5, "WriteCapacityUnits": 5},
                }
            ],
            "BillingMode": "PAY_PER_REQUEST",
        },
    ]

    for table_def in tables:
        name = table_def["TableName"]
        if name in existing:
            print(f"  [ok] {name} already exists")
            continue
        # Remove GSI ProvisionedThroughput for PAY_PER_REQUEST tables
        if table_def.get("BillingMode") == "PAY_PER_REQUEST":
            for gsi in table_def.get("GlobalSecondaryIndexes", []):
                gsi.pop("ProvisionedThroughput", None)
        client.create_table(**table_def)
        print(f"  [ok] Created {name}")

    print("All tables ready.")


# ---------------------------------------------------------------------------
# DynamoDB resource + helper utilities
# ---------------------------------------------------------------------------

def _serialize(obj):
    """Convert Python types to DynamoDB-safe types (float → Decimal)."""
    if isinstance(obj, float):
        return Decimal(str(obj))
    if isinstance(obj, dict):
        return {k: _serialize(v) for k, v in obj.items()}
    if isinstance(obj, list):
        return [_serialize(i) for i in obj]
    return obj


def _deserialize(obj):
    """Convert DynamoDB Decimal back to float."""
    if isinstance(obj, Decimal):
        return float(obj)
    if isinstance(obj, dict):
        return {k: _deserialize(v) for k, v in obj.items()}
    if isinstance(obj, list):
        return [_deserialize(i) for i in obj]
    return obj


def get_dynamodb_resource():
    """Return a boto3 DynamoDB resource, wired to local endpoint if set."""
    kwargs = {"region_name": REGION}
    if ENDPOINT:
        kwargs["endpoint_url"] = ENDPOINT
        kwargs["aws_access_key_id"] = "local"
        kwargs["aws_secret_access_key"] = "local"
    return boto3.resource("dynamodb", **kwargs)


def get_run(run_id: str) -> dict | None:
    """Fetch a single run record by run_id."""
    db = get_dynamodb_resource()
    table = db.Table("finops_runs")
    resp = table.get_item(Key={"run_id": run_id, "sk": "METADATA"})
    item = resp.get("Item")
    return _deserialize(item) if item else None


def get_latest_run() -> dict | None:
    """Return the most recently started run (any status) via full table scan."""
    from boto3.dynamodb.types import TypeDeserializer

    client = get_client()
    resp = client.scan(TableName="finops_runs")
    raw_items = resp.get("Items", [])
    if not raw_items:
        return None

    deserializer = TypeDeserializer()
    items = [
        _deserialize({k: deserializer.deserialize(v) for k, v in item.items()})
        for item in raw_items
    ]
    return max(items, key=lambda x: x.get("started_at", ""))


def get_latest_completed_run() -> dict | None:
    """Query status-started_at-index GSI for the most recent completed run."""
    from boto3.dynamodb.types import TypeDeserializer

    client = get_client()
    resp = client.query(
        TableName="finops_runs",
        IndexName="status-started_at-index",
        KeyConditionExpression="#s = :s",
        ExpressionAttributeNames={"#s": "status"},
        ExpressionAttributeValues={":s": {"S": "completed"}},
        ScanIndexForward=False,
        Limit=1,
    )
    items = resp.get("Items", [])
    if not items:
        return None

    deserializer = TypeDeserializer()
    item = {k: deserializer.deserialize(v) for k, v in items[0].items()}
    return _deserialize(item)


def create_run(
    run_id: str,
    trigger_type: str,
    triggered_by: str,
    okta_token: str,
    tenants_total: int,
) -> None:
    """Insert a new run record with status='running'."""
    db = get_dynamodb_resource()
    table = db.Table("finops_runs")
    now = datetime.now(timezone.utc).isoformat()
    table.put_item(Item=_serialize({
        "run_id": run_id,
        "sk": "METADATA",
        "trigger_type": trigger_type,
        "triggered_by": triggered_by,
        "okta_token": okta_token,
        "status": "running",
        "started_at": now,
        "completed_at": None,
        "tenants_total": tenants_total,
        "tenants_done": 0,
        "hosts_total": 0,
        "hosts_done": 0,
        "log": [],
    }))


def update_run_status(run_id: str, status: str, completed_at: str | None = None) -> None:
    """Update status (and optionally completed_at) for an existing run."""
    db = get_dynamodb_resource()
    table = db.Table("finops_runs")
    expr = "SET #s = :s"
    names = {"#s": "status"}
    values = {":s": status}
    if completed_at:
        expr += ", completed_at = :ca"
        values[":ca"] = completed_at
    table.update_item(
        Key={"run_id": run_id, "sk": "METADATA"},
        UpdateExpression=expr,
        ExpressionAttributeNames=names,
        ExpressionAttributeValues=values,
    )


def get_org_summaries_for_run(run_id: str) -> list[dict]:
    """Query finops_org_summary by run_id via run_id-index GSI."""
    from boto3.dynamodb.types import TypeDeserializer

    client = get_client()
    resp = client.query(
        TableName="finops_org_summary",
        IndexName="run_id-index",
        KeyConditionExpression="run_id = :r",
        ExpressionAttributeValues={":r": {"S": run_id}},
    )
    deserializer = TypeDeserializer()
    items = [
        _deserialize({k: deserializer.deserialize(v) for k, v in item.items()})
        for item in resp.get("Items", [])
    ]
    return items


def get_host_results_for_run(run_id: str) -> list[dict]:
    """Query finops_host_results by run_id via run_id-index GSI.

    The SK is a composite 'host_id#run_id' — split it so callers see just host_id.
    """
    from boto3.dynamodb.types import TypeDeserializer

    client = get_client()
    resp = client.query(
        TableName="finops_host_results",
        IndexName="run_id-index",
        KeyConditionExpression="run_id = :r",
        ExpressionAttributeValues={":r": {"S": run_id}},
    )
    deserializer = TypeDeserializer()
    results = [
        _deserialize({k: deserializer.deserialize(v) for k, v in item.items()})
        for item in resp.get("Items", [])
    ]
    # Split composite SK host_id#run_id → host_id
    for r in results:
        sk = r.get("sk", "")
        if "#" in sk:
            r["host_id"] = sk.split("#")[0]
    return results


if __name__ == "__main__":
    create_tables()

File: packages/agent/src/tools/dynamodb.ts
import { fileURLToPath } from "url";
import { dirname, join } from "path";
import dotenv from "dotenv";
import { DynamoDBClient } from "@aws-sdk/client-dynamodb";
import {
  DynamoDBDocumentClient,
  GetCommand,
  PutCommand,
  UpdateCommand,
  QueryCommand,
  ScanCommand,
} from "@aws-sdk/lib-dynamodb";

// Load env vars — safe to call multiple times (dotenv is idempotent)
const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);
dotenv.config({ path: join(__dirname, "../../.env.local") });

let _client: DynamoDBDocumentClient | null = null;

function getClient(): DynamoDBDocumentClient {
  if (_client) return _client;
  const endpoint = process.env.DYNAMODB_ENDPOINT ?? "";
  const raw = new DynamoDBClient(
    endpoint
      ? { endpoint, region: "us-east-1", credentials: { accessKeyId: "local", secretAccessKey: "local" } }
      : { region: process.env.AWS_REGION ?? "us-east-1" }
  );
  _client = DynamoDBDocumentClient.from(raw);
  return _client;
}

// ─── Host Lists ───────────────────────────────────────────────────────────────

export async function writeHostList(
  tenantId: string,
  runId: string,
  hosts: Array<{ host_id: string; host_name: string }>
): Promise<void> {
  const ttl = Math.floor(Date.now() / 1000) + 7 * 86400;
  await getClient().send(new PutCommand({
    TableName: "finops_host_lists",
    Item: { tenant_id: tenantId, run_id: runId, hosts, ttl },
  }));
}

export async function readHostList(
  tenantId: string,
  runId: string
): Promise<Array<{ host_id: string; host_name: string }>> {
  const resp = await getClient().send(new GetCommand({
    TableName: "finops_host_lists",
    Key: { tenant_id: tenantId, run_id: runId },
  }));
  return (resp.Item?.hosts as Array<{ host_id: string; host_name: string }>) ?? [];
}

// ─── Run Progress ─────────────────────────────────────────────────────────────

export async function updateHostsTotal(runId: string, count: number): Promise<void> {
  await getClient().send(new UpdateCommand({
    TableName: "finops_runs",
    Key: { run_id: runId, sk: "METADATA" },
    UpdateExpression: "SET hosts_total = hosts_total + :n",
    ExpressionAttributeValues: { ":n": count },
  }));
}

export async function updateRunProgress(
  runId: string,
  _tenantId: string,
  hostsDoneIncrement: number,
  logMessage: string
): Promise<void> {
  // Append log entry and increment hosts_done
  await getClient().send(new UpdateCommand({
    TableName: "finops_runs",
    Key: { run_id: runId, sk: "METADATA" },
    UpdateExpression:
      "SET hosts_done = hosts_done + :inc, #log = list_append(if_not_exists(#log, :empty), :msg)",
    ExpressionAttributeNames: { "#log": "log" },
    ExpressionAttributeValues: {
      ":inc": hostsDoneIncrement,
      ":msg": [logMessage],
      ":empty": [],
    },
  }));
  // Trim log to last 20
  const resp = await getClient().send(new GetCommand({
    TableName: "finops_runs",
    Key: { run_id: runId, sk: "METADATA" },
    ProjectionExpression: "#log",
    ExpressionAttributeNames: { "#log": "log" },
  }));
  const log: string[] = (resp.Item?.log as string[]) ?? [];
  if (log.length > 20) {
    await getClient().send(new UpdateCommand({
      TableName: "finops_runs",
      Key: { run_id: runId, sk: "METADATA" },
      UpdateExpression: "SET #log = :trimmed",
      ExpressionAttributeNames: { "#log": "log" },
      ExpressionAttributeValues: { ":trimmed": log.slice(-20) },
    }));
  }
}

export async function updateTenantsDone(runId: string): Promise<void> {
  await getClient().send(new UpdateCommand({
    TableName: "finops_runs",
    Key: { run_id: runId, sk: "METADATA" },
    UpdateExpression: "SET tenants_done = tenants_done + :one",
    ExpressionAttributeValues: { ":one": 1 },
  }));
}

// ─── Host Results ─────────────────────────────────────────────────────────────

export async function writeHostResult(
  tenantId: string,
  runId: string,
  hostId: string,
  result: Record<string, unknown>
): Promise<void> {
  const ttl = Math.floor(Date.now() / 1000) + 90 * 86400;
  await getClient().send(new PutCommand({
    TableName: "finops_host_results",
    Item: {
      tenant_id: tenantId,
      sk: `${hostId}#${runId}`,
      run_id: runId,
      host_id: hostId,
      ttl,
      ...result,
    },
  }));
}

export async function readOrgHostResults(
  tenantId: string,
  runId: string
): Promise<Record<string, unknown>[]> {
  const resp = await getClient().send(new QueryCommand({
    TableName: "finops_host_results",
    IndexName: "run_id-index",
    KeyConditionExpression: "run_id = :r",
    ExpressionAttributeValues: { ":r": runId },
  }));
  const items = (resp.Items ?? []) as Record<string, unknown>[];
  return items.filter((item) => item["tenant_id"] === tenantId);
}

// ─── Org Summary ──────────────────────────────────────────────────────────────

export async function writeOrgSummary(
  tenantId: string,
  runId: string,
  summary: Record<string, unknown>
): Promise<void> {
  const ttl = Math.floor(Date.now() / 1000) + 90 * 86400;
  await getClient().send(new PutCommand({
    TableName: "finops_org_summary",
    Item: { tenant_id: tenantId, run_id: runId, ttl, ...summary },
  }));
}

// ─── Run Status ───────────────────────────────────────────────────────────────

export async function updateRunStatus(
  runId: string,
  status: string,
  completedAt: string
): Promise<void> {
  await getClient().send(new UpdateCommand({
    TableName: "finops_runs",
    Key: { run_id: runId, sk: "METADATA" },
    UpdateExpression: "SET #s = :s, completed_at = :c",
    ExpressionAttributeNames: { "#s": "status" },
    ExpressionAttributeValues: { ":s": status, ":c": completedAt },
  }));
}

// ─── Run Management (API layer) ───────────────────────────────────────────────

export async function createRun(params: {
  runId: string;
  triggerType: string;
  triggeredBy: string;
  oktaToken: string;
  tenantsTotal: number;
}): Promise<void> {
  const now = new Date().toISOString();
  await getClient().send(new PutCommand({
    TableName: "finops_runs",
    Item: {
      run_id: params.runId,
      sk: "METADATA",
      trigger_type: params.triggerType,
      triggered_by: params.triggeredBy,
      okta_token: params.oktaToken,
      status: "running",
      started_at: now,
      completed_at: null,
      tenants_total: params.tenantsTotal,
      tenants_done: 0,
      hosts_total: 0,
      hosts_done: 0,
      log: [],
    },
  }));
}

export async function getRun(runId: string): Promise<Record<string, unknown> | null> {
  const resp = await getClient().send(new GetCommand({
    TableName: "finops_runs",
    Key: { run_id: runId, sk: "METADATA" },
  }));
  return (resp.Item as Record<string, unknown>) ?? null;
}

export async function getLatestRun(): Promise<Record<string, unknown> | null> {
  const resp = await getClient().send(new ScanCommand({
    TableName: "finops_runs",
    FilterExpression: "sk = :sk",
    ExpressionAttributeValues: { ":sk": "METADATA" },
  }));
  const items = (resp.Items ?? []) as Record<string, unknown>[];
  if (!items.length) return null;
  return items.sort((a, b) =>
    String(b["started_at"] ?? "").localeCompare(String(a["started_at"] ?? ""))
  )[0];
}

export async function getLatestCompletedRun(): Promise<Record<string, unknown> | null> {
  const resp = await getClient().send(new QueryCommand({
    TableName: "finops_runs",
    IndexName: "status-started_at-index",
    KeyConditionExpression: "#s = :s",
    ExpressionAttributeNames: { "#s": "status" },
    ExpressionAttributeValues: { ":s": "completed" },
    ScanIndexForward: false,
    Limit: 1,
  }));
  const items = (resp.Items ?? []) as Record<string, unknown>[];
  return items[0] ?? null;
}

export async function getOrgSummariesForRun(runId: string): Promise<Record<string, unknown>[]> {
  const resp = await getClient().send(new QueryCommand({
    TableName: "finops_org_summary",
    IndexName: "run_id-index",
    KeyConditionExpression: "run_id = :r",
    ExpressionAttributeValues: { ":r": runId },
  }));
  return (resp.Items ?? []) as Record<string, unknown>[];
}

export async function getHostResultsForRun(runId: string): Promise<Record<string, unknown>[]> {
  const resp = await getClient().send(new QueryCommand({
    TableName: "finops_host_results",
    IndexName: "run_id-index",
    KeyConditionExpression: "run_id = :r",
    ExpressionAttributeValues: { ":r": runId },
  }));
  const items = (resp.Items ?? []) as Record<string, unknown>[];
  // Split composite SK "host_id#run_id" → expose host_id field
  return items.map((item) => {
    const sk = String(item["sk"] ?? "");
    if (sk.includes("#")) {
      return { ...item, host_id: sk.split("#")[0] };
    }
    return item;
  });
}
